File: 2023/day1.py
import re
import requests
import sys
from pathlib import Path

numbers = {
    "zero": "0",
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9",
}

rnumbers = {k[::-1]: v for k, v in numbers.items()}


# def get_inputs() -> str:
#     response = requests.get("https://adventofcode.com/2023/day/1/input")
#     print(response.status_code)
#     if response.status_code != requests.codes.ok:
#         sys.exit("Can't get inputs")
#     return response.text


# def get_inputs() -> str:
#     return """two1nine
# eightwo
# abcone2threexyz
# xtwone3four
# 4nineeightseven2
# zoneight234
# 7pqrstsixteen
# """


def get_inputs() -> str:
    path = Path("/home/mabdelaa/Downloads/day1-input.txt")
    return path.read_text()


def digit1(line: str) -> str:
    line = re.sub(f'({"|".join(numbers)})', lambda o: numbers[o.group(0)], line, 1)
    return next(c for c in line if c.isdigit())


def digit2(line: str) -> str:
    line = re.sub(f'({"|".join(rnumbers)})', lambda o: rnumbers[o.group(0)], line[::-1], 1)
    return next(c for c in line if c.isdigit())


def convert(line: str) -> int:
    d1 = digit1(line)
    d2 = digit2(line)
    return int(f"{d1}{d2}")


def main():
    content = get_inputs()
    calibration = sum(convert(line) for line in content.splitlines())
    print(calibration)


if __name__ == "__main__":
    main()

# 56042

# 54766
# 55362
# 55360
# 55358

File: 2023/day1.dart
import 'common.dart';

late final List<String> data;

final example1 = r"""
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
""".trim();

final example2 = r"""
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen""".trim();

final numbers = {
    "zero": "0",
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9",
};

final digitRe = RegExp(r"\d");
final numberRe = RegExp('(?=(${numbers.keys.join("|")}))');

List<RegExpMatch> firstLastMatches(Iterable<RegExpMatch> matches) => switch (matches.toList()) {
  [] => [],
  final items => [items.first, items.last],
};

List<RegExpMatch> firstLastDigit(String line) => firstLastMatches(digitRe.allMatches(line));
List<RegExpMatch> firstLastWord(String line) => firstLastMatches(numberRe.allMatches(line));

String convert1(String line) => firstLastDigit(line).map((m) => m[0]!).join();

String convert2(String line) {
  final digitMatches = firstLastDigit(line);
  final wordMatches = firstLastWord(line);

  if (digitMatches.isEmpty)
    return wordMatches.map((m) => numbers[m[1]!]).join();

  if (wordMatches.isEmpty)
    return digitMatches.map((m) =>m[0]!).join();

  final List<String> matches = [
    digitMatches.first.start < wordMatches.first.start
        ? digitMatches.first[0]!
        : numbers[wordMatches.first[1]!]!,
    digitMatches.last.end > wordMatches.last.end
        ? digitMatches.last[0]!
        : numbers[wordMatches.last[1]!]!,
  ];

  return matches.join();
}

Future<void> main() async {
  data = await getData().toList();
  // data = example1.split("\n");
  // data = example2.split("\n");

  final part1 = data.map(convert1).map(int.parse).sum();
  print(part1);

  final part2 = data.map(convert2).map(int.parse).sum();
  print(part2);
}
